libraries/persistent_store/src/format.rs

```diff
@@ -0,0 +1,907 @@
+}
+
+/// Information for an initialized page.
+pub struct InitInfo {
+    /// The number of times this page has been erased.
+    pub cycle: usize,
+
+    /// The word index of the first entry in this virtual page.
+    pub prefix: usize,
+}
+
+/// Information for a page being compacted.
+pub struct CompactInfo {
+    /// The distance in words between head and tail at compaction.
+    pub tail: usize,
+}
+
+/// The first word of an entry.
+#[derive(Debug)]
+pub enum ParsedWord {
+    /// Padding entry.
+    Padding(Padding),
+
+    /// Header of a user entry.
+    Header(Header),
+
+    /// Internal entry.
+    Internal(Internal),
+}
+
+/// Padding entry.
+#[derive(Debug)]
+pub struct Padding {
+    /// The number of following padding words after the first word of the padding entry.
+    pub length: usize,
+}
+
+/// Header of a user entry.
+#[derive(Debug)]
+pub struct Header {
+    /// Whether the last bit of the user data is flipped.
+    pub flipped: bool,
+
+    /// The length in bytes of the user data.
+    pub length: usize,
+
+    /// The key of the user entry.
+    pub key: usize,
+
+    /// The checksum of the user entry.
+    pub checksum: usize,
+}
+
+impl Header {
+    /// Checks the validity of a user entry.
+    ///
+    /// If the user entry has no payload, the `footer` must be set to `None`. Otherwise it should be
+    /// the last word of the entry.
+    pub fn check(&self, footer: Option<&[u8]>) -> bool {
+        footer.map_or(0, |x| count_zeros(x)) == self.checksum
+    }
+}
+
+/// Internal entry.
+#[derive(Debug)]
+pub enum Internal {
+    /// Indicates that a page should be erased.
+    Erase {
+        /// The page to be erased.
+        page: usize,
+    },
+
+    /// Indicates that user entries with high key should be deleted.
+    Clear {
+        /// The minimum key a user entry should have to be deleted.
+        min_key: usize,
+    },
+
+    /// Marks the start of a transaction.
+    ///
+    /// The marker is followed by a given number of updates, which are either user entries or remove
+    /// entries.
+    Marker {
+        /// The number of updates in the transaction.
+        count: usize,
+    },
+
+    /// Indicates that a user entry should be removed.
+    ///
+    /// This is only useful (and valid) as part of a transaction, since removing a single entry is
+    /// already atomic.
+    Remove {
+        /// The key of the user entry to be removed.
+        key: usize,
+    },
+}
+
+/// Returns whether a slice has all bits equal to one.
+pub fn is_erased(slice: &[u8]) -> bool {
+    slice.iter().all(|&x| x == 0xff)
+}
+
+/// Converts a word slice into a word.
+///
+/// # Panics
+///
+/// Panics if `word.len() != WORD_SIZE`.
+fn slice_to_word(word: &[u8]) -> WORD {
+    u32::from_ne_bytes(<[u8; WORD_SIZE]>::try_from(word).unwrap())
+}
+
+/// Divides then takes ceiling.
+///
+/// Returns `ceil(x / m)` with mathematical notations.
+pub const fn div_ceil(x: usize, m: usize) -> usize {
+    (x + m - 1) / m
```

Comment:
Maybe it doesn't break memory safety rules of Rust (e.g. vulnerabilities that expose to arbitrary code execution), but can still break correctness, which for a security key project can be as dramatic.

We can imagine that a bug here could mean reading the wrong entry in storage, which wouldn't break Rust's memory safety rules, but could yield to leaking a secret key for example. There's always a risk that debug mode & fuzzing don't catch a bug happening in production, for lack of coverage.

Reply:
I don't understand what you suggest. Could you clarify?

Reply:
> I don't understand what you suggest. Could you clarify?

Some suggestions:
- Using `checked_add` even in release mode to make sure invariants hold.
- Fixing the arithmetic to be correct even if the addition overflows.
- If the arguments are always `u16`, why not have a `div_ceil(x: u16, m: u16) -> usize` function?

Reply:
- `checked_add`: What to do about the error case that never happens? unwrap? In that case I would suggest turning all arithmetic overflow checks in OpenSK (or just this library) with `overflow-checks`: https://doc.rust-lang.org/cargo/reference/profiles.html#overflow-checks.
- Fix arithmetic: It's adding useless complexity and dead code
- Use `u16`: This simply pushes the problem to the caller

I agree having too much invariant in code is bad. But having simple invariant is needed and helps readability by reducing complexity.

Reply:
> * `checked_add`: What to do about the error case that never happens? unwrap? In that case I would suggest turning all arithmetic overflow checks in OpenSK (or just this library) with `overflow-checks`: https://doc.rust-lang.org/cargo/reference/profiles.html#overflow-checks.

If it doesn't impact performance too much, why not turning on `overflow-checks`? After all, array indexing also panics.

> * Fix arithmetic: It's adding useless complexity and dead code

Likewise, bounds checks in array accesses are useless complexity and dead code.

> * Use `u16`: This simply pushes the problem to the caller

I don't see the problem with that, if indeed the caller guarantees that no overflow happens?

For now, I see the `build_user` function calling `div_ceil` with the length of a slice as argument. This slice is itself provided as a parameter to the `build_user` function, but (1) nothing in the documentation of `build_user` specifies any constraint on the length of this slice and (2) nothing in the `build_user` function checks that the overflow doesn't happen.

Due to (1) and (2) and because `build_user` is a public function, there's nothing guaranteeing that the pre-requisites indeed hold.

By the way, I don't see how the statement that `This function is only called with "small" values (below u16)` holds, in the context of this slice length use case precisely.

> I agree having too much invariant in code is bad. But having simple invariant is needed and helps readability by reducing complexity.

Pre-requisites need to be properly documented though.

---

By the way, the problem of rounded-up division is not new (although there's no solution to it yet): https://github.com/rust-lang/rfcs/issues/2844.

Reply:
> If it doesn't impact performance too much, why not turning on overflow-checks

That's an option (and essentially what I was suggesting). This is orthogonal to this PR though. So you can open an issue if you think it's worth it. I don't think this is more a risk here than it would be in the crypto library for example. This code has very good fuzzing coverage. So we should do something similar for crypto and maybe OpenSK too.

> Likewise, bounds checks in array accesses are useless complexity and dead code

I'm talking about source code complexity, not binary complexity.

> I don't see the problem with that, if indeed the caller guarantees that no overflow happens?

If you don't see a problem with that, then you also don't see a problem with the current code. Because the caller guarantees that no overflow happens.

(1) Do embedded systems have 4G of RAM?
(2) It's a static property (consequence of (1) above) that the overflow does not happen.

More generally `bytes_to_words`, as the name suggests (in addition to the documentation), is used to convert lengths of bytes to words. As such it is called with reasonable inputs because length of bytes can't be arbitrary high, in particular in embedded systems but also due to the storage requirements (page size, erase cycles, etc) that are checked prior to anything else.

> Pre-requisites need to be properly documented though

This is already the case.